OData/src/System.Web.Http.OData/OData/Query/Expressions/SelectExpandBinder.cs

```diff
@@ -115,32 +115,30 @@ internal Expression CreatePropertyNameExpression(IEdmEntityType elementType, IEd
             IEdmEntityType declaringType = property.DeclaringType as IEdmEntityType;
             Contract.Assert(declaringType != null, "only entity types are projected.");
 
-            Expression propertyName;
-
             // derived navigation property using cast
             if (elementType != declaringType)
             {
+                Type originalType = EdmLibHelpers.GetClrType(elementType, _model);
                 Type castType = EdmLibHelpers.GetClrType(declaringType, _model);
                 if (castType == null)
                 {
                     throw new ODataException(Error.Format(SRResources.MappingDoesNotContainEntityType, declaringType.FullName()));
                 }
 
-                // Expression
-                //          source is navigationPropertyDeclaringType ? propertyName : null
-                propertyName = Expression.Condition(
-                    test: Expression.TypeIs(source, castType),
-                    ifTrue: Expression.Constant(property.Name),
-                    ifFalse: Expression.Constant(null, typeof(string)));
-            }
-            else
-            {
-                // Expression
-                //          "propertyName"
-                propertyName = Expression.Constant(property.Name);
+                if (!castType.IsAssignableFrom(originalType))
+                {
+                    // Expression
+                    //          source is navigationPropertyDeclaringType ? propertyName : null
+                    return Expression.Condition(
+                        test: Expression.TypeIs(source, castType),
+                        ifTrue: Expression.Constant(property.Name),
+                        ifFalse: Expression.Constant(null, typeof(string)));
+                }
```

Comment:
Is it possible that castType.IsAssignableFrom(originalType) is false, and Expression.TypeIs(source, castType) is true?